Revit_Core_Engine/Convert/Environment/ToRevit/CableTray.cs

```diff
@@ -0,0 +1,111 @@
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using Autodesk.Revit.DB;
+using Autodesk.Revit.DB.Mechanical;
+using BH.Engine.Adapters.Revit;
+using BH.oM.Adapters.Revit.Settings;
+using BH.oM.Spatial.ShapeProfiles;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using BH.oM.MEP.System.SectionProperties;
+
+namespace BH.Revit.Engine.Core
+{
+    public static partial class Convert
+    {
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        // ToDo for method functionality from top to bottom: 
+        // SectionProperty.ToRevitElementType to BuiltInCategory is unknown and currently set to OST_CableTrayRun
+
+        public static Autodesk.Revit.DB.Electrical.CableTray ToRevitCableTray(this oM.MEP.System.CableTray cableTray, Document document, RevitSettings settings = null, Dictionary<Guid, List<int>> refObjects = null)
+        {
+            // Check valid pipe object
+            if (cableTray == null)
+                return null;
+
+            // Construct Revit Cable Tray
+            Autodesk.Revit.DB.Electrical.CableTray revitTray = refObjects.GetValue<Autodesk.Revit.DB.Electrical.CableTray>(document, cableTray.BHoM_Guid);
+            if (revitTray != null)
+                return revitTray;
+
+            // Settings
+            settings = settings.DefaultIfNull();
+
+            // CableTray type
+            Autodesk.Revit.DB.Electrical.CableTrayType trayType = cableTray.SectionProperty.ToRevitElementType(document, new List<BuiltInCategory> { BuiltInCategory.OST_CableTrayRun }, settings, refObjects) as Autodesk.Revit.DB.Electrical.CableTrayType;
+
+            // End points
+            XYZ start = cableTray.StartPoint.ToRevit();
+            XYZ end = cableTray.EndPoint.ToRevit();
+
+            // Level
+            Level level = document.LevelBelow(Math.Min(start.Z, end.Z), settings);
+            if (level == null)
+                return null;
+
+            revitTray = Autodesk.Revit.DB.Electrical.CableTray.Create(document, trayType.Id, start, end, level.Id);
```

Comment:
This should happen near the end of the method - otherwise we can end up in a situation where the element gets created, and then you throw an error and return null. But the element is already in the model!